### PR TITLE
retroArchCores: remove, retroarchFull: init, retroarch: use fixed paths on "libretro_info_path"

### PR DESCRIPTION
**Pull request by thiagokokada**
###### Motivation for this change

`retroArchCores` is strange: it requires a global configuration on nixpkgs, as:

```nix
nixpkgs.config.retroarch = {
  enableDolphin = true;
  enableMGBA = true;
  enableMAME = true;
};
```

To do so, we ended up declaring all available emulators on `all-packages.nix`. Failing to do so would mean that the emulator wouldn't be available.

However, there is a mechanism on nixpkgs that also works: overrides. Overrides are similar on how other packages works, for example:

```nix
(retroarch.override { cores = with libretro; [ citra snes9x ]; });
```

So let's remove `retroArchCores` and leave the overrides mechanism instead.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by thiagokokada** [comment]
Maybe this needs an entry on release notes :thinking: ?

**Comment by thiagokokada** [comment]
CC @aanderse.

BTW, this doesn't invalidate the need of `programs.retroarch` module. It is basically just a clean-up because I think `retroArchCores` is strange as a mechanism to add cores to packages.

File: pkgs/misc/emulators/retroarch/kodi-advanced-launchers.nix
Comment (thiagokokada) [inline pkgs/misc/emulators/retroarch/kodi-advanced-launchers.nix:3]: Needed to remove the assertion since it would evaluate the assertion before the override occurs.

But I tested and this works fine too:

```nix
    (kodi-retroarch-advanced-launchers.override {
      cores = with pkgs.libretro; [
        snes9x
        desmume
      ];
    })
```

File: pkgs/top-level/all-packages.nix
Comment (thiagokokada) [inline pkgs/top-level/all-packages.nix:32783]: Maybe we could also include a `retroarchFull` derivation that is `retroarch` + all available `libretro` cores. This would make it easier for newbies who are starting on Nix on how to get a working RetroArch setup.

WDYT :thinking: ?

**Comment by aanderse** [comment]
I'm not sure about this. If I run `retroarch` for the first time then `retroarch.cfg` will be created in my `XDG_CONFIG_HOME` directory with the correct values for `libretro_directory` and `libretro_info_path` (because the values will be copied in from ${pkgs.retroarch}/etc/retroarch.cfg`) - but as soon as I do a nixpkgs update + garbage collection then both `libretro_directory` and `libretro_info_path` will be pointing to a path that doesn't exist. I think we **need** a module now.

**Comment by thiagokokada** [comment]
> I'm not sure about this. If I run `retroarch` for the first time then `retroarch.cfg` will be created in my `XDG_CONFIG_HOME` directory with the correct values for `libretro_directory` and `libretro_info_path` (because the values will be copied in from `${pkgs.retroarch}/etc/retroarch.cfg`) - but as soon as I do a nixpkgs update + garbage collection then both `libretro_directory` and `libretro_info_path` will be pointing to a path that doesn't exist. I think we **need** a module now.

No because when the path doesn't exist RetroArch will default to the default config (that is now up-to-date).

Of course, not ideal, but works :man_shrugging: .

BTW, this is not related to this PR right :thinking: ?

**Comment by aanderse** [comment]
> No because when the path doesn't exist RetroArch will default to the default config (that is now up-to-date).

Ah, sorry. I missed that part.

So... should we be consistent and do `retroarch.withPackages (p: with p; [ ... ])` instead of `retroarch.override { cores = with pkgs.libretro; [ ... ] }`? :thinking:

**Comment by aanderse** [comment]
> Maybe this needs an entry on release notes thinking ?

Yes!

**Comment by thiagokokada** [comment]
> So... should we be consistent and do `retroarch.withPackages (p: with p; [ ... ])` instead of `retroarch.override { cores = with pkgs.libretro; [ ... ] }`? thinking

I mean, we could, but I still prefer the module because the situation is not ideal (for example, if I added/remove a core, until I trigger the GC this will reflect the old cores from the old derivation).

However, this is orthogonal with this PR. Even with the `programs.retroarch` we will need to support non-NixOS systems, e.g.: macOS. In this case the only option they have is to use either `override` or `withPackages`, even if it is sub-optiomal.

So my proposal is to use `overrides` AND add `programs.retroarch`. However I think the module can come in another PR.

**Comment by thiagokokada** [comment]
> However, this is orthogonal with this PR. Even with the programs.retroarch we will need to support non-NixOS systems, e.g.: macOS. In this case the only option they have is to use either override or withPackages, even if it is sub-optiomal

BTW, I want to eventually solve this problem, however I am trying to ask upstream for help so we don't need to patch too much. See issue https://github.com/libretro/RetroArch/issues/13251 for details.

Even if a better solution for the core issue came afterwards, I still think that having a module is worth it. With it, we can allow folks to fully customize RetroArch (for example, this would make NixOS a good replacement for RetroPie).

File: pkgs/top-level/all-packages.nix
Comment (thiagokokada) [inline pkgs/top-level/all-packages.nix:29201]: I just discovered that this one didn't even work (the correct name is `thepowedertoy`).

**Comment by thiagokokada** [comment]
> Yes!

Done.

Also added a release notes entry about the recent RetroArch bump.

**Comment by thiagokokada** [comment]
Result of `nixpkgs-review pr 146714` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>kodi-retroarch-advanced-launchers</li>
    <li>retroarchFull</li>
  </ul>
</details>

**Comment by thiagokokada** [comment]
@aanderse Can I have your review?

**Comment by thiagokokada** [comment]
> BTW, I want to eventually solve this problem, however I am trying to ask upstream for help so we don't need to patch too much. See issue [libretro/RetroArch#13251](https://github.com/libretro/RetroArch/issues/13251) for details.

With commit f69afeb, I finally solved this issue. Now RetroArch completely ignores the value of `libretro_info_path` (it already ignored `libretro_directory`) is set on `retroarch.cfg`. Better yet, it always updates the value on the configuration file, so this is less confusing to the users.

WDYT @aanderse ? I think this solves most of the issues. I can also split this in another PR if you want.

**Review by aanderse — APPROVED** [review]
This looks great! It all makes sense to me, and the release notes cover everything we need to. Thanks for doing all of this 🙇

**Comment by thiagokokada** [comment]
> This looks great! It all makes sense to me, and the release notes cover everything we need to. Thanks for doing all of this bow

Nice.

I will just test this on macOS since I need to patch one more place to make it work.

**Comment by thiagokokada** [comment]
@aanderse Sorry for bothering you, but can you do one last review with the Darwin changes?

File: pkgs/misc/emulators/retroarch/default.nix
Comment (aanderse) [inline pkgs/misc/emulators/retroarch/default.nix:115]: I just saw a ton of people volunteer to support OSX recently. I wonder if we could find some volunteers after we merge this?

**Comment by aanderse** [comment]
No bother at all. Looks good to me 👍

File: pkgs/misc/emulators/retroarch/default.nix
Comment (thiagokokada) [reply pkgs/misc/emulators/retroarch/default.nix:115]: Yeah, this is my idea.

BTW, it is already broken on `master`, because Wayland is Linux-only (so it propagates its `broken` state).

Comment (thiagokokada) [reply pkgs/misc/emulators/retroarch/default.nix:115]: On `master`:

```
nix-build -A retroarch
error: Package ‘wayland-1.19.0’ in /Users/thiagoko/Projects/nixpkgs/pkgs/development/libraries/wayland/default.nix:107 is marked as broken, refusing to evaluate.

       a) To temporarily allow broken packages, you can use an environment variable
          for a single invocation of the nix tools.

            $ export NIXPKGS_ALLOW_BROKEN=1

       b) For `nixos-rebuild` you can set
         { nixpkgs.config.allowBroken = true; }
       in configuration.nix to override this.

       c) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
         { allowBroken = true; }
       to ~/.config/nixpkgs/config.nix.
(use '--show-trace' to show detailed location information)
```

**Comment by thiagokokada** [comment]
> No bother at all. Looks good to me +1

Nice, I will do one last build to make sure everything is fine and merge them.

**Comment by aanderse** [comment]
Great. Thanks again! I'll try to create a maintainers team PR this weekend.

I'm really loving to `libretro` integration in `kodi` 😄

File: pkgs/misc/emulators/retroarch/0001-Disable-menu_show_core_updater.patch
Comment (matthewbauer) [inline pkgs/misc/emulators/retroarch/0001-Disable-menu_show_core_updater.patch:19]: Does this disable manually installed cores too? I don’t know if anyone actually does this but I thought in the past you could imperitively install cores in the UI. If this can be enabled manually in the UI that would be alright.

Comment (thiagokokada) [reply pkgs/misc/emulators/retroarch/0001-Disable-menu_show_core_updater.patch:19]: No, it just disable the UI to download new cores, that didn't work anyway since it doesn't connect to the RetroArch core servers (it would just show the cores that you already have installed). I don't know if this used to work, but at least the situation from version 1.8.5 (the version that was on `master` before I started updating and refactor everything) was already like this.

You can still re-enable this manually though, and this is respected on the next boot. But it will be useless.

File: pkgs/misc/emulators/retroarch/0002-Use-fixed-paths-on-libretro_info_path.patch
Comment (matthewbauer) [inline pkgs/misc/emulators/retroarch/0002-Use-fixed-paths-on-libretro_info_path.patch:63]: Could we set the DEFAULT_DIR_CORE constant instead? If I understand this, it is assuming the default is unset?

Comment (thiagokokada) [reply pkgs/misc/emulators/retroarch/0002-Use-fixed-paths-on-libretro_info_path.patch:63]: No, this is just a enum. This is exactly the part of the code that sets this enum to the default.

I tried to set it before but got either a compilation failure or segfault (don't remember exactly).

File: pkgs/misc/emulators/retroarch/0001-Disable-menu_show_core_updater.patch
Comment (thiagokokada) [reply pkgs/misc/emulators/retroarch/0001-Disable-menu_show_core_updater.patch:19]: BTW, even if this used to work I don't think download would do anything.

First because this is NixOS, and thanks to this a core downloaded from RetroArch servers would probably just segfault (each core is a `.so` file and are dynamically linked, however since NixOS has no FHS it wouldn't find the necessary libraries).

Second because even before this PR, `libretro_directory` was already set to a path on `/nix/store`, that is read-only, so it would fail to write the core there. I am just setting `libretro_directory` directly in the code for consistency, but it was already the default.

**Comment by thiagokokada** [comment]
@matthewbauer Any other issue? Can I go and merge this (still building everything, but the only core that remains are MAME and should finish soon).

**Comment by thiagokokada** [comment]
Saw this PR that should be included from MAME 0.225 and above (we use 0.227 currently), so I think it is safe to enable parallel building in everything, except for older MAME cores: https://github.com/mamedev/mame/pull/7279.

Restart my build from zero (hope this is faster though, building the latest version of MAME in special is very slow without parallel building).

**Comment by thiagokokada** [comment]
> Saw this PR that should be included from MAME 0.225 and above (we use 0.227 currently), so I think it is safe to enable parallel building in everything, except for older MAME cores: [mamedev/mame#7279](https://github.com/mamedev/mame/pull/7279).
> 
> Restart my build from zero (hope this is faster though, building the latest version of MAME in special is very slow without parallel building).

Seems like it was a good idea, it build much faster:

Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>82 packages built:</summary>
  <ul>
    <li>kodi-retroarch-advanced-launchers (xbmc-retroarch-advanced-launchers)</li>
    <li>kodiPackages.libretro-genplus</li>
    <li>kodiPackages.libretro-mgba</li>
    <li>kodiPackages.libretro-snes9x</li>
    <li>libretro.atari800</li>
    <li>libretro.beetle-gba</li>
    <li>libretro.beetle-lynx</li>
    <li>libretro.beetle-ngp</li>
    <li>libretro.beetle-pce-fast</li>
    <li>libretro.beetle-pcfx</li>
    <li>libretro.beetle-psx</li>
    <li>libretro.beetle-psx-hw</li>
    <li>libretro.beetle-saturn</li>
    <li>libretro.beetle-saturn-hw</li>
    <li>libretro.beetle-snes</li>
    <li>libretro.beetle-supergrafx</li>
    <li>libretro.beetle-vb</li>
    <li>libretro.beetle-wswan</li>
    <li>libretro.bluemsx</li>
    <li>libretro.bsnes-mercury</li>
    <li>libretro.citra</li>
    <li>libretro.desmume</li>
    <li>libretro.desmume2015</li>
    <li>libretro.dolphin</li>
    <li>libretro.dosbox</li>
    <li>libretro.eightyone</li>
    <li>libretro.fbalpha2012</li>
    <li>libretro.fbneo</li>
    <li>libretro.fceumm</li>
    <li>libretro.flycast</li>
    <li>libretro.fmsx</li>
    <li>libretro.freeintv</li>
    <li>libretro.gambatte</li>
    <li>libretro.genesis-plus-gx</li>
    <li>libretro.gpsp</li>
    <li>libretro.gw</li>
    <li>libretro.handy</li>
    <li>libretro.hatari</li>
    <li>libretro.mame</li>
    <li>libretro.mame2000</li>
    <li>libretro.mame2003</li>
    <li>libretro.mame2003-plus</li>
    <li>libretro.mame2010</li>
    <li>libretro.mame2015</li>
    <li>libretro.mame2016</li>
    <li>libretro.mesen</li>
    <li>libretro.meteor</li>
    <li>libretro.mgba</li>
    <li>libretro.mupen64plus</li>
    <li>libretro.neocd</li>
    <li>libretro.nestopia</li>
    <li>libretro.np2kai</li>
    <li>libretro.o2em</li>
    <li>libretro.opera</li>
    <li>libretro.parallel-n64</li>
    <li>libretro.pcsx_rearmed</li>
    <li>libretro.picodrive</li>
    <li>libretro.play</li>
    <li>libretro.ppsspp</li>
    <li>libretro.prboom</li>
    <li>libretro.prosystem</li>
    <li>libretro.quicknes</li>
    <li>libretro.sameboy</li>
    <li>libretro.scummvm</li>
    <li>libretro.smsplus-gx</li>
    <li>libretro.snes9x</li>
    <li>libretro.snes9x2002</li>
    <li>libretro.snes9x2005</li>
    <li>libretro.snes9x2010</li>
    <li>libretro.stella</li>
    <li>libretro.stella2014</li>
    <li>libretro.tgbdual</li>
    <li>libretro.thepowdertoy</li>
    <li>libretro.tic80</li>
    <li>libretro.vba-m</li>
    <li>libretro.vba-next</li>
    <li>libretro.vecx</li>
    <li>libretro.virtualjaguar</li>
    <li>libretro.yabause</li>
    <li>retroarch</li>
    <li>retroarchBare</li>
    <li>retroarchFull</li>
  </ul>
</details>

**Pull request merged**
